### PR TITLE
Add support for complex path drawing

### PR DESCRIPTION
**Pull request by birktj**
This PR adds primitives for more complex path drawing.

Everything is added inside a `draw` module.

The supported operations are currently:
- `close`
- `moveto`
- `lineto`
- `cubicto`
- `arcto`
- `circle`

However it should be relatively easy to add new operations in the future.

This allows for code like this:
```
#draw.path(stroke: blue, fill: blue.lighten(80%), {
    import draw: *
    moveto(y: 10pt)
    for ms in (1, -1) {
        for sxy in ((1, -1), (-1, 1)) {
            let sx = sxy.at(0)
            let sy = sxy.at(1)
            let i = 0
            while i < 4 {
                lineto(dx: ms*sx*10pt, dy: ms*sy*10pt)
                lineto(dx: ms*10pt, dy: ms*10pt)
                i += 1
            }
        }
    }
    close()
    circle(x: 40pt, y: 50pt, r: 15pt, ccw: false)
})
```
image

**Ideas for future development:**
- Add transformation operations for paths (translate, rotate, etc..). I experimented with this, but had some trouble with delta/absolute coordinates as well as relative lengths so I skipped it for now.
- Add support for using path as a clipping mask

Related to #260

### CODE REVIEW

File: src/geom/path_builder.rs [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+    }
+}
+
+/// A value that is either absolute or a delta of an
+/// earlier value.
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub enum DeltaAbs<T> {
+    Abs(T),
+    Delta(T),
+}
```
Comment (birktj) [inline src/geom/path_builder.rs:272]: I am not completely in love with this enum yet, so better suggestions for name or an entirely different strategy would be welcome.

File: library/src/visualize/draw.rs [hunk 2]
```diff
@@ -0,0 +1,521 @@
...
+///         lineto(dx: ms*sx*10pt, dy: ms*sy*10pt)
+///         lineto(dx: ms*10pt, dy: ms*10pt)
+///         i += 1
+///       }
+///     }
+///   }
+///   close()
+///   circle(x: 40pt, y: 50pt, r: 15pt, ccw: false)
+/// })
+/// ```
```
Comment (birktj) [inline library/src/visualize/draw.rs:48]: Maybe this example is a bit much?

**Comment by PgBiel** [comment]
This is an awesome idea - I like how you used the Typst-way:tm: of joining a sequence of things by simply returning them (or, rather, implementing the add operator, and that property follows).

If combined with a `node` instruction and, maybe eventually, `canvas` block... Maybe we will have some preliminary TikZ-like thing going on here? :eyes:

File: library/src/visualize/draw.rs [hunk 3]
```diff
@@ -0,0 +1,521 @@
...
+    /// ```example
+    /// #draw.path(stroke: blue, fill: blue.lighten(80%), {
+    ///   draw.moveto(x: 20pt, y: 20pt)
+    ///   draw.circle(r: 20pt)
+    ///   draw.circle(r: 10pt, ccw: true)
+    /// })
+    /// ```
+    #[named]
+    #[default]
+    ccw: bool,
```
Comment (giovanniberti) [inline library/src/visualize/draw.rs:481]: How about renaming `ccw` to `counterclockwise`? (or, even shorter, invert the logic and name it `clockwise`)

**Comment by laurmaedje** [comment]
I'm not really a fan of having two path systems. Maybe this should supersede the current path function? Would also like to hear @SekoiaTree's opinion as they contributed the current `path` function.

On an unrelated note, I'm thinking about adding scopes to functions, so that this could just be:
```
#path(stroke: blue, fill: blue.lighten(80%), {
  path.moveto(y: 10pt)
  ..
})
```
And the same with image effects.

**Comment by PgBiel** [comment]
> Maybe this should supersede the current path function?

Could we maybe just use the current path system as a shorthand for several `moveto` and/or curve instructions then? If we manage to unify them in this manner, this could reduce maintenance overhead overall if this gets implemented.

> On an unrelated note, I'm thinking about adding scopes to functions

Ohh, that's really cool. That could be really useful for `canvas`, as we could use a local scope for things like (using some sample, arbitrary syntax here) `node("name", x: 2pt, y: 3pt)[a]`, or like `at(2pt, 3pt)[a]` or similar, without polluting the global namespace, while also not having to rely on "context parameters".

Could this also work e.g. when specifying `[   #path.moveto(...)  ]` (content)  instead of code?

**Comment by SekoiaTree** [comment]
If I'm being honest, the current path syntax is really quite ugly and a pain to work with. The best thing I can say about it is that it works. I'm 100% fine with it being superseded.

This is much better, but I'm not a fan of the implicit-ish state that gets added? I think I'd prefer some kind of chained operator thing, where every method takes itself and returns a new object, and that's immutable. Essentially expose PathBuilder as a type to the user.

**Comment by laurmaedje** [comment]
> This is much better, but I'm not a fan of the implicit-ish state that gets added?

It's not really implicit, is it? It just makes use of Typst's joining semantics.

**Comment by SekoiaTree** [comment]
Yeah, true. Still, I think I'd prefer a type, though.

**Comment by Dherse** [comment]
I think that custom join semantic could be useful in multiple areas, image effect would also benefit from this. And I am sure that a few other places might benefit from it. But what if we restricted it further my making it something along the lines of:

```
#path(stroke: blue, fill: blue.lighten(80%), path => {
  path.moveto(y: 10pt)
  ..
})
```

Where instead of being an implicit context, we make it explicit with a closure? That way the context doesn't pollute the main context of typst and isn't useful outside of the `path` anyway and it is nicely contained and made explicit.

**Comment by laurmaedje** [comment]
> That way the context doesn't pollute the main context of typst

It's possible, but it makes it harder to automatically generate documentation. I outlined an alternative approach [here](https://github.com/typst/typst/pull/673#issuecomment-1503342253).

File: library/src/visualize/draw.rs (hunk 3 above)
Comment (birktj) [reply library/src/visualize/draw.rs:481]: Arcs already have a `sweep` parameter which is the same thing (taken from [svg](https://developer.mozilla.org/en-US/docs/Web/SVG/Tutorial/Paths#arcs)). They should probably at least be consistent. I don't know if `sweep` is much clearer though? `counterclockwise` and `clockwise` are at least both quite clear, but maybe a bit verbose?

**Comment by fenjalien** [comment]
Would it make more sense to have a `canvas` element? Johannes has created a poc implementation in pure typst, theres a demo here: https://typst.app/project/wEVFkcvIVnJeQyFDlAhfHC. and a thread here: https://discord.com/channels/1054443721975922748/1097035369641410621

**Comment by johannes-wolf** [comment]
… on GitHub: https://github.com/johannes-wolf/typst-canvas

**Comment by PgBiel** [comment]
> Would it make more sense to have a `canvas` element?

I think both ideas aren't exclusive; canvas will also need some form of path drawing. However, of course, the two things should interact very well with each other in the future.

**Comment by jason-s** [comment]
Is there a spec / design doc somewhere? I don't grok Rust, but I think you ought to specify the behavior first before merging.

I love the general idea. Which model is this aimed after? SVG? HTML Canvas? something else?

**Comment by AlistairKeiller** [comment]
> Is there a spec / design doc somewhere? I don't grok Rust, but I think you ought to specify the behavior first before merging.
> 
> I love the general idea. Which model is this aimed after? SVG? HTML Canvas? something else?

One existing design that was [mentioned in the Discord](https://discord.com/channels/1054443721975922748/1056599302912360518/1085926393260548097), I really like, and aligns with the philosophy of a immutable type to represent a diagram is [diagrammar](https://www.youtube.com/watch?v=gT9Xu-ctNqI). It is an internal tool at brilliant, so that video is the only resource on it, but it is a interesting video with many interesting ideas.

**Comment by SekoiaTree** [comment]
The mentioned discord conversation is not available to anyone who wasn't part of the preview :(

**Comment by laurmaedje** [comment]
I'd like to wait a bit here to:
(a) work out the function scope idea (#843)
(b) consider if and how this interacts with a potential canvas primitive (which also needs to integrate with other drawing primitives like circles, etc.)

**Comment by birktj** [comment]
That sounds good. Just let me know when you are ready to move this forward and what work you want be to do.

The way I personally see it is that the main contribution of this PR is the `path` type itself, not so much the actual path drawing (filling and stroking) which is the obvious (but not only) application of such a path type. For me a goal is also to create something similar to [paths in the _basic layer_ of PGF](https://tikz.dev/base-paths).

I imagine it should be possible to extend such a path type with functions like:
- Shorten or extend a path at either end
- Get the point at a certain distance along the path
- Get the direction of a path at a certain distance along it
- Compute the bounding box of a path
- Apply transformations to a path

**Comment by johannes-wolf** [comment]
Any updates on this?

**Comment by laurmaedje** [comment]
> Any updates on this?

Unfortunately, not really. There is so much to work on that this part of the language is advancing very slowly. 

However, I can see that blocking this indefinitely isn't really helpful. Maybe we can find a way to do this is in a minimally useful way to keep breakage minimal should we rework it in the future.

**Comment by laurmaedje** [comment]
I'm closing this for now. We should to go back to the drawing board once work on the new layout engine starts and custom types have landed.

**Comment by Enivex** [comment]
> I'm closing this for now. We should to go back to the drawing board once work on the new layout engine starts and custom types have landed.

Ba dum tss

**Pull request closed without merging**